diff --git a/build/pdfmake.js b/build/pdfmake.js
index 7b00f1c4679073c50d0d17f25973443d0b1d0835..ef9ad9f8ab62c6785b34f9bb69292f50ec79beba 100644
--- a/build/pdfmake.js
+++ b/build/pdfmake.js
@@ -47099,7 +47099,7 @@ var SVGtoPDF = function SVGtoPDF(doc, svg, x, y, options) {
     doc.addContent('ET');
   }
   function docFillColor(color) {
-    if (color[0].constructor.name === 'PDFPattern') {
+    if (color[0].constructor.name === 'PDFPattern' || color[0].dx) {
       doc.fillOpacity(color[1]);
       docUsePattern(color[0], false);
     } else {
@@ -47107,7 +47107,7 @@ var SVGtoPDF = function SVGtoPDF(doc, svg, x, y, options) {
     }
   }
   function docStrokeColor(color) {
-    if (color[0].constructor.name === 'PDFPattern') {
+    if (color[0].constructor.name === 'PDFPattern' || color[0].dx) {
       doc.strokeOpacity(color[1]);
       docUsePattern(color[0], true);
     } else {
