diff --git a/node_modules/pdfmake/build/pdfmake.js b/node_modules/pdfmake/build/pdfmake.js
index 7547d87..0d90240 100644
--- a/node_modules/pdfmake/build/pdfmake.js
+++ b/node_modules/pdfmake/build/pdfmake.js
@@ -20345,7 +20345,7 @@ var SVGtoPDF = function SVGtoPDF(doc, svg, x, y, options) {
     doc.addContent('ET');
   }
   function docFillColor(color) {
-    if (color[0].constructor.name === 'PDFPattern') {
+    if (color[0].constructor.name === 'PDFPattern' || color[0].dx) {
       doc.fillOpacity(color[1]);
       docUsePattern(color[0], false);
     } else {
@@ -20353,7 +20353,7 @@ var SVGtoPDF = function SVGtoPDF(doc, svg, x, y, options) {
     }
   }
   function docStrokeColor(color) {
-    if (color[0].constructor.name === 'PDFPattern') {
+    if (color[0].constructor.name === 'PDFPattern' || color[0].dx) {
       doc.strokeOpacity(color[1]);
       docUsePattern(color[0], true);
     } else {